xonsh/commands_cache.py

```diff
@@ -184,9 +183,62 @@ def predict_threadable(self, cmd):
                 pre, ext = os.path.splitext(name)
                 if pre in predictors:
                     predictors[name] = predictors[pre]
+        if name not in predictors:
+            predictors[name] = self.default_predictor(name, cmd[0])
         predictor = predictors[name]
         return predictor(cmd[1:])
 
+    #
+    # Background Predictors (as methods)
+    #
+
+    def default_predictor(self, name, cmd0):
+        if ON_POSIX:
+            return self.default_predictor_readbin(name, cmd0,
+                                                  timeout=.1,
+                                                  failure=predict_true)
+        else:
+            return predict_true
+
+    def default_predictor_readbin(self, name, cmd0, timeout, failure):
+        """Make a defautt predictor by
+        analyzing the content of the binary. Should only works on POSIX.
+        Return failure if the analysis fails.
+        """
+
+        fname = cmd0 if os.path.isabs(cmd0) else self.locate_binary(name)
```

Comment:
This should probably call `lazy_locate_binary()` instead. The cache should be filled by SubprocSpec by the time we get to the prediction stage.